lib/expect.h

```diff
@@ -0,0 +1,126 @@
+// limitations under the License.
+
+#ifndef EXPECT_H_
+#define EXPECT_H_
+
+#include <complex>
+
+#include "fuser.h"
+#include "gate_appl.h"
+
+namespace qsim {
+
+template <typename Gate>
+struct OpString {
+  std::complex<double> weight;
+  std::vector<Gate> ops;
+};
+
+/**
+ * Computes the expectation value of the sum of operator strings. Operators
+ * can be any supported gates. Uses a temporary state vector.
+ * @param param Options for gate fusion.
+ * @param strings Operator strings.
+ * @param ss StateSpace object required to copy the state vector and compute
+ *   inner products.
+ * @param simulator Simulator object. Provides specific implementations for
+ *   applying gates.
+ * @param state The state vector of the system.
+ * @param state Temporary state vector.
+ * @return The computed expectation value.
+ */
+template <typename Fuser, typename Simulator, typename Gate>
+std::complex<double> ExpectaionValue(const typename Fuser::Parameter& param,
+                                     const std::vector<OpString<Gate>>& strings,
+                                     const typename Simulator::StateSpace& ss,
+                                     const Simulator& simulator,
+                                     const typename Simulator::State& state,
+                                     typename Simulator::State& ket) {
+  std::complex<double> eval = 0;
+
+  for (const auto& str : strings) {
+    if (str.ops.size() == 0) continue;
+
+    ss.Copy(state, ket);
+
+    if (str.ops.size() == 1) {
+      const auto& op = str.ops[0];
+      simulator.ApplyGate(op.qubits, op.matrix.data(), ket);
+    } else {
+      auto fused_gates = Fuser::FuseGates(param, state.num_qubits(), str.ops);
+      if (fused_gates.size() == 0) {
+        eval = 0;
+        break;
+      }
+
+      for (const auto& fgate : fused_gates) {
+        ApplyFusedGate(simulator, fgate, ket);
+      }
+    }
+
+    eval += str.weight * ss.InnerProduct(state, ket);
+  }
+
+  return eval;
+}
+
+/**
+ * Computes the expectation value of the sum of operator strings. Operators
+ * can be any supported gates except for user-defined controlled gates.
+ * Computation is performed in place. No additional memory is allocated. The
+ * operator strings should act on no more than six qubits and they should be
+ * fusible into one gate.
+ * @param strings Operator strings.
+ * @param simulator Simulator object. Provides specific implementations for
+ *   computing expectation values.
+ * @param state The state of the system.
+ * @return The computed expectation value.
+ */
+template <typename Fuser, typename Simulator, typename Gate>
+std::complex<double> ExpectaionValue(const std::vector<OpString<Gate>>& strings,
+                                     const Simulator& simulator,
+                                     const typename Simulator::State& state) {
+  std::complex<double> eval = 0;
+
+  typename Fuser::Parameter param;
+  param.max_fused_size = 6;
```

Comment:
Would calculating `max_fused_size` from the number of qubits in `strings` be worthwhile here?

Reply:
It does not affect anything if `max_fused_size` is greater than the actual number of qubits.